src/ai/backend/storage/vast/vastdata_client.py

```diff
@@ -0,0 +1,440 @@
+        body: dict[str, Any] = {
+            "name": str(path),
+            "path": str(self.storage_base_dir / path.name),
+            "create_dir": False,  # Explicitly disable to create directory owned by root.
+        }
+        if soft_limit is not None:
+            body["soft_limit"] = soft_limit
+        if hard_limit is not None:
+            body["hard_limit"] = hard_limit
+        if soft_limit_inodes is not None:
+            body["soft_limit_inodes"] = soft_limit_inodes
+        if hard_limit_inodes is not None:
+            body["hard_limit_inodes"] = hard_limit_inodes
+        if grace_period is not None:
+            body["grace_period"] = grace_period
+
+        async with aiohttp.ClientSession(
+            base_url=self.api_endpoint,
+        ) as sess:
+            response = await self._build_request(
+                sess,
+                POST,
+                "quotas/",
+                body,
+            )
+            data: Mapping[str, Any] = await response.json()
+            match response.status:
+                case 201 | 200:
+                    pass
+                case 400 | 401:
+                    raise VastInvalidParameterError
+                case 403:
+                    raise VastUnauthorizedError
+                case 503:
+                    raise QuotaScopeAlreadyExists
+                case _:
+                    raise VastUnknownError(
+                        f"Unkwon error from vast API. status code: {response.status}"
+                    )
+        return VastQuota.from_json(data)
+
+    async def modify_quota(
+        self,
+        vast_quota_id: VastQuotaID,
+        soft_limit: int | None = None,
+        hard_limit: int | None = None,
+        soft_limit_inodes: int | None = None,
+        hard_limit_inodes: int | None = None,
+        grace_period: str | None = None,
+    ) -> VastQuota:
+        body: dict[str, Any] = {}
+        if soft_limit is not None:
+            body["soft_limit"] = soft_limit
+        if hard_limit is not None:
+            body["hard_limit"] = hard_limit
+        if soft_limit_inodes is not None:
+            body["soft_limit_inodes"] = soft_limit_inodes
+        if hard_limit_inodes is not None:
+            body["hard_limit_inodes"] = hard_limit_inodes
+        if grace_period is not None:
+            body["grace_period"] = grace_period
+
+        async with aiohttp.ClientSession(
+            base_url=self.api_endpoint,
+        ) as sess:
+            response = await self._build_request(
+                sess,
+                PATCH,
+                f"quotas/{vast_quota_id}/",
+                body,
+            )
+            data: Mapping[str, Any] = await response.json()
+            match response.status // 100:
+                case 2:
+                    pass
+                case 4 | 5:
+                    err_msg = data.get("detail", "Unkown error from vast API")
+                    raise VastInvalidParameterError(err_msg)
```

Comment:
In theory `Invalid Parameter` error should only be represented by error code 400, not whole 4xx and 5xx status codes. We should discuss whether this kind of approach is legitimate or not.

Reply:
I updated this part as handling 4xx and 5xx in the same way would be buggy.
The reason why I handled them in the same way is that some invalid API call with wrong parameter or invalid condition among parameters (such as, setting soft limit bigger than hard limit) would responses with status code 500 but it is client's fault.